Text as Image node is broken for PIL 10

@adegerard Do you happen to know why this happens?


![image](https://github.com/chaiNNer-org/chaiNNer/assets/20878432/d29d0afd-0da1-4477-90cb-9ec03a3de4e8)
![image](https://github.com/chaiNNer-org/chaiNNer/assets/20878432/54a2d9cb-9572-4cbf-b6dc-0370bcc7df9e)


Logs:

<details>

```
[2023-09-13 20:14:00 +0200] [6492] [INFO] Sanic v23.3.0
[2023-09-13 20:14:00 +0200] [6492] [INFO] Goin' Fast @ http://127.0.0.1:8000
[2023-09-13 20:14:00 +0200] [6492] [INFO] mode: production, single worker
[2023-09-13 20:14:00 +0200] [6492] [INFO] server: sanic, HTTP/1.1
[2023-09-13 20:14:00 +0200] [6492] [INFO] python: 3.8.2
[2023-09-13 20:14:00 +0200] [6492] [INFO] platform: Windows-10-10.0.19041-SP0
[2023-09-13 20:14:00 +0200] [6492] [INFO] packages: sanic-routing==22.8.0
[2023-09-13 20:14:00 +0200] [6492] [INFO] Starting setup...
[2023-09-13 20:14:00 +0200] [6492] [INFO] Starting worker [6492]
[2023-09-13 20:14:00 +0200] [6492] [INFO] Importing nodes...
[2023-09-13 20:14:00 +0200] [6492] [INFO] Checking dependencies...
[2023-09-13 20:14:00 +0200] [6492] [INFO] Checking dependencies for chaiNNer_standard...
[2023-09-13 20:14:00 +0200] [6492] [INFO] Checking dependencies for PyTorch...
[2023-09-13 20:14:00 +0200] [6492] [INFO] Checking dependencies for NCNN...
[2023-09-13 20:14:00 +0200] [6492] [INFO] Checking dependencies for ONNX...
[2023-09-13 20:14:00 +0200] [6492] [INFO] Checking dependencies for External...
[2023-09-13 20:14:00 +0200] [6492] [INFO] Done checking dependencies...
[2023-09-13 20:14:02 +0200] [6492] [INFO] Sending backend ready...
[2023-09-13 20:14:02 +0200] [6492] [INFO] Done.
[2023-09-13 21:00:10 +0200] [6492] [INFO] Running new executor...
[2023-09-13 21:00:10 +0200] [6492] [INFO] Execution options: {}
[2023-09-13 21:00:10 +0200] [6492] [ERROR] 'FreeTypeFont' object has no attribute 'getsize'
Traceback (most recent call last):
  File "C:\Users\micha\Git\chaiNNer\backend\src\process.py", line 106, in run_node
    raw_output = node.run(*enforced_inputs)
  File "C:\Users\micha\Git\chaiNNer\backend\src\packages\chaiNNer_standard\image\create_images\text_as_image.py", line 173, in text_as_image_node
    w_ref, h_ref = font.getsize(max_line)[0], font.getsize("[�]")[1]
AttributeError: 'FreeTypeFont' object has no attribute 'getsize'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\Users\micha\Git\chaiNNer\backend\src\server.py", line 189, in run
    await executor.run()
  File "C:\Users\micha\Git\chaiNNer\backend\src\process.py", line 576, in run
    await self.__process_nodes(self.__get_output_nodes())
  File "C:\Users\micha\Git\chaiNNer\backend\src\process.py", line 563, in __process_nodes
    await self.process(output_node)
  File "C:\Users\micha\Git\chaiNNer\backend\src\process.py", line 389, in process
    return await self.__process(node)
  File "C:\Users\micha\Git\chaiNNer\backend\src\process.py", line 418, in __process
    processed_input = await self.process(node_input.id)
  File "C:\Users\micha\Git\chaiNNer\backend\src\process.py", line 389, in process
    return await self.__process(node)
  File "C:\Users\micha\Git\chaiNNer\backend\src\process.py", line 444, in __process
    output, execution_time = await self.loop.run_in_executor(
  File "C:\Python38\lib\concurrent\futures\thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "C:\Users\micha\Git\chaiNNer\backend\src\process.py", line 325, in wrapper
    result = supplier()
  File "C:\Users\micha\Git\chaiNNer\backend\src\process.py", line 115, in run_node
    raise NodeExecutionError(node_id, node, str(e), input_dict) from e
process.NodeExecutionError: 'FreeTypeFont' object has no attribute 'getsize'
[2023-09-13 21:00:10 +0200] [6492] [ERROR] Traceback (most recent call last):
  File "C:\Users\micha\Git\chaiNNer\backend\src\process.py", line 106, in run_node
    raw_output = node.run(*enforced_inputs)
  File "C:\Users\micha\Git\chaiNNer\backend\src\packages\chaiNNer_standard\image\create_images\text_as_image.py", line 173, in text_as_image_node
    w_ref, h_ref = font.getsize(max_line)[0], font.getsize("[�]")[1]
AttributeError: 'FreeTypeFont' object has no attribute 'getsize'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\Users\micha\Git\chaiNNer\backend\src\server.py", line 189, in run
    await executor.run()
  File "C:\Users\micha\Git\chaiNNer\backend\src\process.py", line 576, in run
    await self.__process_nodes(self.__get_output_nodes())
  File "C:\Users\micha\Git\chaiNNer\backend\src\process.py", line 563, in __process_nodes
    await self.process(output_node)
  File "C:\Users\micha\Git\chaiNNer\backend\src\process.py", line 389, in process
    return await self.__process(node)
  File "C:\Users\micha\Git\chaiNNer\backend\src\process.py", line 418, in __process
    processed_input = await self.process(node_input.id)
  File "C:\Users\micha\Git\chaiNNer\backend\src\process.py", line 389, in process
    return await self.__process(node)
  File "C:\Users\micha\Git\chaiNNer\backend\src\process.py", line 444, in __process
    output, execution_time = await self.loop.run_in_executor(
  File "C:\Python38\lib\concurrent\futures\thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "C:\Users\micha\Git\chaiNNer\backend\src\process.py", line 325, in wrapper
    result = supplier()
  File "C:\Users\micha\Git\chaiNNer\backend\src\process.py", line 115, in run_node
    raise NodeExecutionError(node_id, node, str(e), input_dict) from e
process.NodeExecutionError: 'FreeTypeFont' object has no attribute 'getsize'
```

</details>

--- a/backend/src/nodes/impl/caption.py
+++ b/backend/src/nodes/impl/caption.py
@@ -1,6 +1,9 @@
+from __future__ import annotations
+
 import os
 import sys
 from enum import Enum
+from typing import Tuple
 
 import cv2
 import numpy as np
@@ -13,6 +16,17 @@ from .image_utils import normalize, to_uint8
 class CaptionPosition(Enum):
     BOTTOM = "bottom"
     TOP = "top"
+
+
+def get_font_size(
+    font: ImageFont.ImageFont | ImageFont.FreeTypeFont, text: str
+) -> Tuple[int, int]:
+    """Get font [width, height] of the given text"""
+    # (left, top, right, bottom)
+    caption_bb = font.getbbox(text)
+    font_width = caption_bb[2] - caption_bb[0]
+    font_height = caption_bb[3] - caption_bb[1]
+    return font_width, font_height
 
 
 def add_caption(
@@ -44,7 +58,7 @@ def add_caption(
         text_y = round(size / 2)
     font_color = (255,) * c
 
-    fw, _ = font.getsize(caption)
+    fw, _ = get_font_size(font, caption)
     # scale font size to fit image
     if fw > w:
         font = ImageFont.truetype(font_path, round(fontsize * w / fw))

--- a/backend/src/nodes/impl/caption.py
+++ b/backend/src/nodes/impl/caption.py
@@ -18,9 +18,7 @@ class CaptionPosition(Enum):
     TOP = "top"
 
 
-def get_font_size(
-    font: ImageFont.ImageFont | ImageFont.FreeTypeFont, text: str
-) -> Tuple[int, int]:
+def get_font_size(font: ImageFont.FreeTypeFont, text: str) -> Tuple[int, int]:
     """Get font [width, height] of the given text"""
     # (left, top, right, bottom)
     caption_bb = font.getbbox(text)

--- a/backend/src/packages/chaiNNer_standard/image/create_images/text_as_image.py
+++ b/backend/src/packages/chaiNNer_standard/image/create_images/text_as_image.py
@@ -7,6 +7,7 @@ from enum import Enum
 import numpy as np
 from PIL import Image, ImageDraw, ImageFont
 
+from nodes.impl.caption import get_font_size
 from nodes.impl.color.color import Color
 from nodes.impl.image_utils import normalize, to_uint8
 from nodes.properties.inputs import (
@@ -170,14 +171,14 @@ def text_as_image_node(
 
     # Use a text as reference to get max size
     font = ImageFont.truetype(font_path, size=100)
-    w_ref, h_ref = font.getsize(max_line)[0], font.getsize("[§]")[1]
+    w_ref, h_ref = get_font_size(font, max_line)[0], get_font_size(font, "[§]")[1]
 
     # Calculate font size to fill the specified image size
     w = int(width * 100.0 / w_ref)
     h = int(height * 100.0 / (h_ref * line_count))
     font_size = min(w, h)
     font = ImageFont.truetype(font_path, size=font_size)
-    w_text, h_text = font.getsize(max_line)
+    w_text, h_text = get_font_size(font, max_line)
     h_text *= line_count
 
     # Text color

